components/imageproc/src/processor.rs

```diff
@@ -68,33 +71,51 @@ impl ImageOp {
+            }
+        };
+
         match self.format {
             Format::Png => {
-                img.write_to(&mut tmp_output_writer, ImageFormat::Png)?;
+                let mut encoder = PngEncoder::new(&mut tmp_output_writer);
+                add_color_profile(&mut encoder);
+                img.write_with_encoder(encoder)?;
             }
             Format::Jpeg { quality } => {
                 let mut encoder = JpegEncoder::new_with_quality(&mut tmp_output_writer, quality);
+                add_color_profile(&mut encoder);
                 encoder.encode_image(&img)?;
             }
             Format::WebP { quality } => {
-                let encoder = webp::Encoder::from_image(&img)
-                    .map_err(|_| anyhow!("Unable to load this kind of image with webp"))?;
-                let memory = match quality {
-                    Some(q) => encoder.encode(q as f32),
-                    None => encoder.encode_lossless(),
-                };
-                tmp_output_writer.write_all(memory.as_bytes())?;
+                // use the `image` builtin encoder for lossless, as it supports color profiles
+                if quality.is_none() && has_color_profile {
+                    let mut encoder = WebPEncoder::new_lossless(&mut tmp_output_writer);
+                    add_color_profile(&mut encoder);
+                    img.write_with_encoder(encoder)?;
+                } else {
+                    if has_color_profile {
+                        println!(
+                            "processing {}: Warning: Lossy WebP encoder does not support color profiles, colors may be incorrect.",
+                            self.input_path.display()
+                        );
+                    }
+                    let encoder = webp::Encoder::from_image(&img)
+                        .map_err(|_| anyhow!("Unable to load this kind of image with webp"))?;
+                    let memory = match quality {
+                        Some(q) => encoder.encode(q as f32),
+                        None => encoder.encode_lossless(),
+                    };
+                    tmp_output_writer.write_all(memory.as_bytes())?;
+                }
             }
             Format::Avif { quality, speed } => {
-                let mut avif: Vec<u8> = Vec::new();
-                let encoder = AvifEncoder::new_with_speed_quality(&mut avif, speed, quality);
-                encoder.write_image(
-                    img.as_bytes(),
-                    img.dimensions().0,
-                    img.dimensions().1,
-                    img.color().into(),
-                )?;
-                tmp_output_writer.write_all(avif.as_bytes())?;
+                let mut encoder =
+                    AvifEncoder::new_with_speed_quality(&mut tmp_output_writer, speed, quality);
+                add_color_profile(&mut encoder);
+                img.write_with_encoder(encoder)?;
```

Comment:
This change does two more things beyond adding the profile:
- it also bypasses the encode to memory
- it removes the img.as_bytes() call, which makes a copy of the whole image

Those two should make the encode faster by 1x output data memcpy + 1x image data memcpy. Since it matches all the other standard encoders, the tests pass and my site still works, I’m confident this is correct.